catalog/core/catalog-core-standardframework/src/test/java/ddf/catalog/cache/solr/impl/CachingFederationStrategyTest.java

```diff
@@ -0,0 +1,383 @@
+                .within(WITHIN_TIMEOUT, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testFederateQueryAndUpdateCacheBlocking() throws Exception {
+        HashMap<String, Serializable> properties = new HashMap<>();
+        properties.put(QUERY_MODE, INDEX_QUERY_MODE);
+
+        QueryRequest fedQueryRequest = new QueryRequestImpl(mockQuery, properties);
+
+        MockSource mockSource = getMockSource(mockResponse);
+
+        List<Source> sources = Arrays.asList(mockSource);
+
+        when(mockSortedQueryMonitorFactory.createMonitor(any(),
+                any(),
+                any(),
+                any(),
+                any())).thenAnswer(invocation -> {
+            Object[] args = invocation.getArguments();
+            QueryResponseImpl queryResponse = (QueryResponseImpl) args[2];
+            queryResponse.setHits(HITS_RESULT);
+            QueryRequestImpl queryRequest = (QueryRequestImpl) args[3];
+            assertThat(queryRequest.getPropertyValue(QUERY_MODE), is(INDEX_QUERY_MODE));
+            return mockSortedQueryMonitor;
+        });
+
+        QueryResponse federateResponse = strategy.federate(sources, fedQueryRequest);
+
+        WaitCondition.expect("Wait until the result appears in the response")
+                .until(() -> federateResponse.getHits(), is(HITS_RESULT))
+                .checkEvery(POLLING_INTERVAL, TimeUnit.MILLISECONDS)
+                .within(WITHIN_TIMEOUT, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testFederateQueryAndUpdateCacheNotBlocking() throws Exception {
+        HashMap<String, Serializable> properties = new HashMap<>();
+        properties.put(QUERY_MODE, UPDATE_QUERY_MODE);
+
+        QueryRequest fedQueryRequest = new QueryRequestImpl(mockQuery, properties);
+
+        MockSource mockSource = getMockSource(mockResponse);
+
+        List<Source> sources = Arrays.asList(mockSource);
+
+        when(mockSortedQueryMonitorFactory.createMonitor(any(),
+                any(),
+                any(),
+                any(),
+                any())).thenAnswer(invocation -> {
+            Object[] args = invocation.getArguments();
+            QueryResponseImpl queryResponse = (QueryResponseImpl) args[2];
+            queryResponse.setHits(HITS_RESULT);
+            QueryRequestImpl queryRequest = (QueryRequestImpl) args[3];
+            assertThat(queryRequest.getPropertyValue(QUERY_MODE), is(UPDATE_QUERY_MODE));
+            return mockSortedQueryMonitor;
+        });
+
+        QueryResponse federateResponse = strategy.federate(sources, fedQueryRequest);
+
+        WaitCondition.expect("Wait until the result appears in the response")
+                .until(() -> federateResponse.getHits(), is(HITS_RESULT))
+                .checkEvery(POLLING_INTERVAL, TimeUnit.MILLISECONDS)
+                .within(WITHIN_TIMEOUT, TimeUnit.SECONDS);
+    }
+
+    @Test
+    public void testFederateQueryCacheWithOffsetHandler() throws Exception {
```

Comment:
An `OffsetResultHandler` is only created if the offset in `federate()` is > 1, but the if-statement in the while loop in `OffsetResultHandler.run()` only passes if the `queryResultIndex` variable in `OffsetResultHandler.run()` is >= offset, but `queryResultIndex` is always set to 1. Maybe I'm missing something, but I don't see how that if-statement could ever pass.